### ISSUE TITLE
Razor compilation APIs exposing Roslyn types being obsoleted in 2.2.0 and removed in 3.0.0

### ISSUE THREAD

**Issue by danroth27** [aspnet/AspNetCore#4902]
As part of the 2.1.0 release, ASP.NET Core added support for build time compilation of cshtml files (Razor Views & Razor Pages). This feature is on by default and allows build time verification of the correctness of Razor files, and also faster startup performance. Additionally, MVC watches for changes to cshtml files and seamlessly updates build-time compiled views with runtime compiled views. This allows editing cshtml files and then refreshing the browser to see the changes.

MVC uses the app's compilation context, details about the references and settings that were used to compile the source code (.cs) in the app, to compile views at runtime. In addition to the compilation context, MVC exposes options such as [RazorViewEngineOptions.CompilationCallback](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.razor.razorviewengineoptions.compilationcallback?view=aspnetcore-2.1) and [AdditionalCompilationReferences](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.razor.razorviewengineoptions.additionalcompilationreferences?view=aspnetcore-2.1) that allow configuring compilation specifically for runtime. 

Runtime specific customizations are unavailable at build time and consequently the two outputs could be different.

To address this inconsistency, we plan on removing APIs in MVC that change or modify runtime compilation. For 2.2.0, we're marking these APIs obsolete with the intent to remove them in 3.0.0. This is the list of affected APIs:

* Properties on `Microsoft.AspNetCore.Mvc.Razor.RazorViewEngineOptions`
    * `AdditionalCompilationReferences`
    * `CompilationCallback`

* `Microsoft.AspNetCore.Mvc.Razor.Compilation.MetadataReferenceFeature`
* `Microsoft.AspNetCore.Mvc.Razor.Compilation.MetadataReferenceFeatureProvider`
* `Microsoft.AspNetCore.Mvc.Razor.Compilation.RazorReferenceManager`

No changes are required to your application if it does not use these APIs.

Apps using these APIs to add assembly references to the compilation context for runtime compilation should instead use [`ApplicationPartManager.AddApplicationPart`](https://docs.microsoft.com/dotnet/api/microsoft.extensions.dependencyinjection.mvccoremvccorebuilderextensions.addapplicationpart) to add application parts for each assembly reference, or switch to a built-time compilation model (see [Create reusable UI using the Razor Class Library project](https://docs.microsoft.com/aspnet/core/razor-pages/ui-class)).

Other scenarios blocked by the removal of these APIs and not covered by alternative MVC features should raised here.

**Comment by ngohungphuc** [comment]
Hi @danroth27 I have below code so how can I change it accordingly to AddApplicationPart because AddApplicationPart require assembly to be pass in
![devenv_mqw62b3mzz](https://user-images.githubusercontent.com/13591213/49984778-15c4cc00-ff9c-11e8-83cf-c8a9dd3d9ee6.png)

**Comment by pranavkm** [comment]
What do you use to create the passed in instance of `MetadataReference`?

**Comment by ngohungphuc** [comment]
@pranavkm I just added my code please check it.
Thank you

**Comment by pranavkm** [comment]
Looks like you're already adding the assembly as an application part on line 115. Removing the `foreach` loop that adds compilation references should suffice.

**Comment by MathiasKowoll** [comment]
I have the same issue, just adding the aplicación part assembly does not register views and models from those compilations.

**Comment by thiennn** [comment]
In https://github.com/simplcommerce/SimplCommerce, with it's modular architecture, each module is a Razor Class Library. We manually load modules and then add them to the main web app using ApplicationPart. 

We also support theming, users can switch between the themes at run time. This feature required run time compilation of cshtml file. Adding ApplicationPart seems not work this scenario

https://github.com/simplcommerce/SimplCommerce/blob/894813fe6f52e5a28307d424deb198686f78110c/src/SimplCommerce.WebHost/Extensions/ServiceCollectionExtensions.cs#L79

**Comment by pranavkm** [comment]
@MathiasKowoll here's the bit of code that's required to register assembly parts including the views assembly: https://github.com/pranavkm/PluginApp/blob/master/MainApp/Startup.cs#L74-L90

> This feature required run time compilation of cshtml file. Adding ApplicationPart seems not work this scenario

@thiennn can you be more precise about what does not work? I'd recommend filing a separate issue with a simple repro to show what's expected and what does not work.

**Comment by camelofcode** [comment]
I have an issue the same as https://github.com/aspnet/AspNetCore/issues/2126

I solved this using @pranavkm 's solution on that issue (Aug 14, 2017).

However, since MetadataReferenceFeature is being obsoleted, I am trying to rewrite so it doesn't use it.

I'm basically doing what @pranavkm 's other link says (https://github.com/pranavkm/PluginApp/blob/master/MainApp/Startup.cs#L74-L90) now, but it makes no difference, always get this error, unless using the workaround in issue 2126.


[14:38:42 ERR] An unhandled exception has occurred while executing the request.
System.InvalidOperationException: Cannot find compilation library location for p
ackage 'Modular.Modules.Application1'
   at Microsoft.Extensions.DependencyModel.CompilationLibrary.ResolveReferencePa
ths(ICompilationAssemblyResolver resolver, List`1 assemblies)
   at Microsoft.Extensions.DependencyModel.CompilationLibrary.ResolveReferencePa
ths()
   at Microsoft.AspNetCore.Mvc.ApplicationParts.AssemblyPart.<>c.<GetReferencePa
ths>b__8_0(CompilationLibrary library)
   at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.MoveNext()
   at Microsoft.AspNetCore.Mvc.Razor.Compilation.MetadataReferenceFeatureProvide
r.PopulateFeature(IEnumerable`1 parts, MetadataReferenceFeature feature)
   at Microsoft.AspNetCore.Mvc.ApplicationParts.ApplicationPartManager.PopulateF
eature[TFeature](TFeature feature)
   at Microsoft.AspNetCore.Mvc.Razor.Internal.DefaultRazorReferenceManager.GetCo
mpilationReferences()
   at System.Threading.LazyInitializer.EnsureInitializedCore[T](T& target, Boole
an& initialized, Object& syncLock, Func`1 valueFactory)
   at Microsoft.AspNetCore.Mvc.Razor.Internal.DefaultRazorReferenceManager.get_C
ompilationReferences()
   at Microsoft.AspNetCore.Mvc.Razor.Internal.LazyMetadataReferenceFeature.get_R
eferences()
   at Microsoft.CodeAnalysis.Razor.CompilationTagHelperFeature.GetDescriptors()
   at Microsoft.AspNetCore.Razor.Language.DefaultRazorTagHelperBinderPhase.Execu
teCore(RazorCodeDocument codeDocument)
   at Microsoft.AspNetCore.Razor.Language.RazorEnginePhaseBase.Execute(RazorCode
Document codeDocument)
   at Microsoft.AspNetCore.Razor.Language.DefaultRazorEngine.Process(RazorCodeDo
cument document)
   at Microsoft.AspNetCore.Razor.Language.DefaultRazorProjectEngine.ProcessCore(
RazorCodeDocument codeDocument)
   at Microsoft.AspNetCore.Razor.Language.RazorProjectEngine.Process(RazorProjec
tItem projectItem)
   at Microsoft.AspNetCore.Mvc.Razor.Internal.RazorViewCompiler.CompileAndEmit(S
tring relativePath)
   at Microsoft.AspNetCore.Mvc.Razor.Internal.RazorViewCompiler.OnCacheMiss(Stri
ng normalizedPath)

A screenshot of Modular.Modules.Application1 sitting in mvcBuilder.PartManager.ApplicationParts, and having the path that it really exists at on disk.
![image](https://user-images.githubusercontent.com/19360404/50907803-255ef580-1420-11e9-9113-beaf67620941.png)

**Comment by MathiasKowoll** [comment]
@camelofcode look at this bug #6215

**Comment by camelofcode** [comment]
@MathiasKowoll Thanks, that worked great - deleting the deps.json file for the dll fixes it. Not a surprise as my code is heavily inspired by @thiennn 's article here https://www.codeproject.com/Articles/1109475/Modular-Web-Application-with-ASP-NET-Core - so anything that fixes his problems likely helps me!

**Comment by MathiasKowoll** [comment]
@camelofcode You are welcome, my code is also heavily bases on @thiennn work.

**Comment by aspnet-hello** [comment]
We periodically close 'discussion' issues that have not been updated in a long period of time.

We apologize if this causes any inconvenience. We ask that if you are still encountering an issue, please log a new issue with updated information and we will investigate.
